# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import cross_val_score, KFold
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
data = pd.read_csv("science_data_large.csv")

# Output the first 15 rows of the data
print("First 15 rows of the dataset:")
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print("\nDataset Information:")
data.info()
print("\nStatistical Summary:")
print(data.describe())

First 15 rows of the dataset:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL       

## 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model

sample_data = pd.DataFrame({
    "Temperature °C": [500],
    "Mols KCL": [300]         
})

predicted_size = model.predict(sample_data)
print("Predicted Size (nm^3):", predicted_size)

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("R² Score:", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)


Predicted Size (nm^3): [333490.24704829]
R² Score: 0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.4795834081


Write the linear equation of a slime: (example equation: $E = mc^2$)

$\hat{y}$ =−409391.48 + 866.15 (Temperature °C) + 1032.70 (Mols KCl)

Report on score and explain meaning: The R² score of 0.855 indicates that approximately 85.5% of the variance in the slime's size is explained by our model using the two features: temperature and moles of KCl. The model has a strong linear relationship with the observed data, capturing most of the variability, while only about 14.5% of the variation remains unexplained. This high R² score suggests that our linear regression model is effective in predicting the growth behavior of the slime under different experimental conditions.

## 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model, X, y, cv=kf)

# Report on their finding and their significance
print("Cross Validation Scores:", cv_scores)
print("Average CV Score:", cv_scores.mean())

Cross Validation Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Average CV Score: 0.8597294202684646


Write findings here:

Fold Scores:
Fold 1: 0.8615
Fold 2: 0.8274
Fold 3: 0.8720
Fold 4: 0.8817
Fold 5: 0.8561
Average Score: 0.8597

Findings and Interpretation

__Consistent Performance:__
The scores across the five folds range from approximately 0.827 to 0.882. This relatively narrow range indicates that the model performs consistently across different subsets of the data, suggesting good generalizability.

__High Variance Explanation:__
With an average R² score of about 0.86, the model explains around 86% of the variability in the slime's size based on temperature and moles of KCl. This reinforces the earlier test score and shows that the linear regression model captures a strong linear relationship.

__Robustness:__
The high and consistent R² values imply that the model is robust. It performs well not just on a single train-test split, but across multiple splits, reducing concerns about overfitting.

## 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

model_poly = LinearRegression()

# Perform k-fold cross validation (as above)
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores_poly = cross_val_score(model_poly, X_poly, y, cv=kf)

# Report on the metrics and output the resultant equation as you did in Part 3.
print("Polynomial Regression Cross Validation Scores:", cv_scores_poly)
print("Average CV Score:", cv_scores_poly.mean())

model_poly.fit(X_poly, y)
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

Polynomial Regression Cross Validation Scores: [1. 1. 1. 1. 1.]
Average CV Score: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Polynomial Regression Equation:

h(x) = 0.00002 + 12.00000(Temperature °C) + -0.00000(Mols KCL) + -0.00000(Temperature °C)^2 + 2.00000(Temperature °C * Mols KCL) + 0.02857(Mols KCL)^2

Report on the score and interpret:

Findings and Interpretation

__Perfect Fit:__
An R² score of 1.0 across all folds indicates that the model is able to perfectly explain the variance in the slime's size based on the features. This means that every data point is predicted without any error.

__Implications on the Data Relationship:__
The perfect fit suggests that the relationship between the experimental conditions and the slime’s size is likely exactly quadratic (or can be perfectly modeled by a second-degree polynomial). The augmented features capture the underlying relationship completely.

__Robustness of the Model:__
Since the cross validation was performed over different splits of the dataset and each fold yielded a perfect score, this is not merely an artifact of overfitting to one particular split. Instead, it indicates that the quadratic model is robust across the dataset.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [7]:

# Load the cleaned CKD dataset
data = pd.read_csv("ckd_feature_subset.csv")

# Display the first few rows to verify the dataset structure
print("First 5 rows of the dataset:")
print(data.head())

# Define features and target variable
X = data.drop("Target_ckd", axis=1)
y = data["Target_ckd"]

# Create a 5-fold cross-validation object for consistency
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Define the classifiers
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "Support Vector Machine": SVC(random_state=42),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(max_iter=1000, random_state=42)
}

# Dictionary to store results
results = {}

for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=kf)
    results[name] = {"Mean Accuracy": scores.mean(), "Std": scores.std()}

# Convert the results to a DataFrame for a clean table display
results_df = pd.DataFrame(results).T
print("Classification Model Performance:")
print(results_df)


First 5 rows of the dataset:
        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
Classification Model Performance:
                        Mean Accuracy       Std
Logistic Regression          0.856559  0.066269
Support Vector Machine       0.928172  0.047601
K-Nearest Neighbors          0.927957  0.052440
Neural Network               0.935054  0.040466


## Results and Conclusion for Classification Experiments

#### Findings and Conclusions 
##### Dataset Overview
__Dataset Structure:__ The CKD dataset contains normalized features such as age, blood pressure (bp), white blood cell count (wbcc), appetite indicators (appet_poor, appet_good), red blood cell count (rbcc), and a binary target variable (Target_ckd). The first five rows indicate the expected structure and normalization of the features.
Classification Models Performance

__Logistic Regression:__

Mean Accuracy: 85.66%

Std Dev: 6.63%

Observation: As a linear classifier, it provided a baseline performance but struggled with capturing more complex, non-linear relationships in the data.

__Support Vector Machine (SVC):__

Mean Accuracy: 92.82%

Std Dev: 4.76%

Observation: SVC improved performance by finding a robust decision boundary in the feature space, indicating non-linear separability in the data.

__K-Nearest Neighbors (KNN):__

Mean Accuracy: 92.80%

Std Dev: 5.24%

Observation: KNN performed comparably to SVC, confirming that local neighborhood information is quite informative for this classification task.

__Neural Network (MLP):__

Mean Accuracy: 93.51%

Std Dev: 4.05%

Observation: The neural network achieved the highest accuracy among the standard classifiers, suggesting that its capacity to model non-linear interactions is beneficial for this dataset.

In [8]:

# Define different neural network configurations
nn_configurations = {
    "NN Default": MLPClassifier(max_iter=1000, random_state=42),
    "NN Two Hidden Layers": MLPClassifier(hidden_layer_sizes=(50, 50), max_iter=3000, random_state=42),
    "NN Tanh Activation": MLPClassifier(activation='tanh', max_iter=1000, random_state=42),
    "NN Higher Learning Rate": MLPClassifier(learning_rate_init=0.01, max_iter=3000, random_state=42)
}

# Dictionary to store NN results
nn_results = {}

for name, model in nn_configurations.items():
    scores = cross_val_score(model, X, y, cv=kf)
    nn_results[name] = {"Mean Accuracy": scores.mean(), "Std": scores.std()}

# Convert the NN results to a DataFrame for display
nn_results_df = pd.DataFrame(nn_results).T
print("Neural Network Configuration Performance:")
print(nn_results_df)


Neural Network Configuration Performance:
                         Mean Accuracy       Std
NN Default                    0.935054  0.040466
NN Two Hidden Layers          0.954624  0.043696
NN Tanh Activation            0.941505  0.037327
NN Higher Learning Rate       0.961075  0.047382


## Results and Conclusion for Neural Network Experiments

#### Neural Network Configurations Experiments

__NN Default:__

Mean Accuracy: 93.51%

Std Dev: 4.05%


__NN Two Hidden Layers:__

Mean Accuracy: 95.46%

Std Dev: 4.37%

Finding: Adding an extra hidden layer enhances the network’s ability to capture complex feature interactions, resulting in improved performance.


__NN Tanh Activation:__

Mean Accuracy: 94.15%

Std Dev: 3.73%

Finding: Changing the activation function to tanh provided a moderate boost in performance, though it was still slightly behind the two-hidden-layer configuration.


__NN Higher Learning Rate:__

Mean Accuracy: 96.11%

Std Dev: 4.74%

Finding: Increasing the learning rate resulted in the best overall performance. This suggests that a more aggressive update allowed the model to converge more effectively to a high-performing solution, although care must be taken as too high a rate could risk instability.


#### Overall Conclusions

__Best Overall Model:__

The neural network with a higher learning rate achieved the highest accuracy (96.11%). This indicates that the dataset benefits from models capable of capturing non-linearities, and that appropriate hyperparameter tuning (such as learning rate adjustment) can lead to significant performance improvements.

__Model Robustness:__
The lower standard deviation for the neural network configurations compared to logistic regression implies that the neural network models are more robust across different data splits, providing consistent performance.

__Implications for CKD Classification:__
These results underscore the importance of using advanced machine learning techniques, particularly neural networks with well-tuned hyperparameters, for the classification of chronic kidney disease.